10_rest/willshen8/cmd/puppy-server/main_test.go

```diff
@@ -0,0 +1,95 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/anz-bank/go-course/10_rest/willshen8/pkg/puppy"
+
+	"github.com/stretchr/testify/assert"
+)
+
+// Unit test for parseCmdArgs() with long flags
+func TestParseCmdArgs(t *testing.T) {
+	args = []string{"--data", "./../../data/puppy.json", "--port", "1234", "--store", "map"}
+	returnedArgs := parseCmdArgs(args)
+	assert.Equal(t, "./../../data/puppy.json", returnedArgs.file)
+	assert.Equal(t, 1234, returnedArgs.port)
+	assert.Equal(t, "map", returnedArgs.storer)
+}
+
+// Unit test for parseCmdArgs() with short flags
+func TestShortFlag(t *testing.T) {
+	args = []string{"-d", "./../../data/puppy.json", "-p", "1234", "-s", "map"}
+	returnedArgs := parseCmdArgs(args)
+	assert.Equal(t, "./../../data/puppy.json", returnedArgs.file)
+	assert.Equal(t, 1234, returnedArgs.port)
+	assert.Equal(t, "map", returnedArgs.storer)
+}
+
+// Unit test (positive test) for parsePortFlag()
+func TestParseCorrectPort(t *testing.T) {
+	// setup the test server with arguments
+	args = []string{"--data", "./../../data/puppy.json", "--port", "1234"}
+	parsedArgs := parseCmdArgs(args)
+	flagReturned, _ := parsePortFlag(parsedArgs)
+	var expectedFlag = "1234"
+	assert.Equal(t, expectedFlag, flagReturned)
+}
+
+// Unit test (negative test) for parsePortFlag()
+func TestParseWrongPort(t *testing.T) {
+	// setup the test server with arguments
+	args = []string{"--data", "./../../data/puppy.json", "--port", "123456789"}
+	parsedArgs := parseCmdArgs(args)
+	flagReturned, flagError := parsePortFlag(parsedArgs)
+	var expectedFlag = "0"
+	var expectedError = "Invalid port number entered."
+	assert.Equal(t, expectedFlag, flagReturned)
+	assert.Equal(t, expectedError, flagError.Error())
+
+}
+
+// Unit test for createPuppies()
+func TestSavePuppiesToStore(t *testing.T) {
+	var buf bytes.Buffer
+	out = &buf
+	var puppies []puppy.Puppy
+	puppies = append(puppies, puppy.Puppy{ID: 1, Breed: "Jack Russell Terrier", Color: "White and Brown", Value: "1500"})
+	storer := puppy.NewMapStore()
+	returnedPuppyStorer, _ := createPuppies(storer, puppies)
+
+	expected := "&{map[1:{1 Jack Russell Terrier White and Brown 1500}] 1}\n"
+	fmt.Fprintln(out, returnedPuppyStorer)
+	actual := buf.String()
+	assert.Equal(t, expected, actual)
+}
+
+func TestMain(m *testing.M) {
+	// setup the test server with arguments
+	args = []string{"--data", "./../../data/puppy.json", "--port", "1234", "--store", "map"}
+	os.Exit(m.Run())
+}
+
+func TestParseCmdFlagsError(t *testing.T) {
+	args = []string{"--wrongFlag"}
+	assert.NotPanics(t, main)
+	os.Exit(0)
```

Comment:
I'm also a bit confused that os.Exit doesn't exit the test run here.
Could you please run your tests with `-v` and show at least here that all tests run please (copy-paste the logs)?

Reply:
```
shenw-JSS1256:willshen8 shenw$ make
go build ./...
go test ./...
ok  	github.com/anz-bank/go-course/10_rest/willshen8/cmd/puppy-server	0.018s
ok  	github.com/anz-bank/go-course/10_rest/willshen8/pkg/puppy	(cached)
golangci-lint run
go test -coverprofile=coverage.out ./... && go tool cover -html=coverage.out
ok  	github.com/anz-bank/go-course/10_rest/willshen8/cmd/puppy-server	0.014s
ok  	github.com/anz-bank/go-course/10_rest/willshen8/pkg/puppy	0.017s	coverage: 100.0% of statements```

Reply:
This function should be:
``func TestParseCmdFlagsError(t *testing.T) {
	args = []string{"--wrongFlag"}
	assert.Panics(t, main)
}

Reply:
```shenw-JSS1256:willshen8 shenw$ go test -v ./...
=== RUN   TestParseCmdArgs
--- FAIL: TestParseCmdArgs (0.00s)
    main_test.go:18:
        	Error Trace:	main_test.go:18
        	Error:      	Not equal:
        	            	expected: string("./../../data/puppy.json")
        	            	actual  : *os.File(&os.File{file:(*os.file)(0xc0000d62a0)})
        	Test:       	TestParseCmdArgs
=== RUN   TestShortFlag
--- FAIL: TestShortFlag (0.00s)
    main_test.go:27:
        	Error Trace:	main_test.go:27
        	Error:      	Not equal:
        	            	expected: string("./../../data/puppy.json")
        	            	actual  : *os.File(&os.File{file:(*os.file)(0xc0000d6360)})
        	Test:       	TestShortFlag
=== RUN   TestParseCorrectPort
--- PASS: TestParseCorrectPort (0.00s)
=== RUN   TestParseWrongPort
--- PASS: TestParseWrongPort (0.00s)
=== RUN   TestSavePuppiesToStore
--- PASS: TestSavePuppiesToStore (0.00s)
=== RUN   TestParseCmdFlagsError
--- PASS: TestParseCmdFlagsError (0.00s)
=== RUN   TestBadFileName
--- PASS: TestBadFileName (0.00s)
=== RUN   TestInvalidJSON
--- PASS: TestInvalidJSON (0.00s)
=== RUN   TestUnmarshalJSON
--- PASS: TestUnmarshalJSON (0.00s)
FAIL
FAIL	github.com/anz-bank/go-course/10_rest/willshen8/cmd/puppy-server	0.013s
=== RUN   TestUnknownError
--- PASS: TestUnknownError (0.00s)
=== RUN   TestRestAPI
=== RUN   TestRestAPI/Test_GET_Puppy_by_ID
=== RUN   TestRestAPI/Test_GET_Puppy_with_invalid_input
=== RUN   TestRestAPI/Test_GET_invalid_Puppy
=== RUN   TestRestAPI/Test_POST_Puppy
=== RUN   TestRestAPI/Test_POST_with_error_data
=== RUN   TestRestAPI/Test_POST_with_decode_error
=== RUN   TestRestAPI/Test_PUT_puppy_by_ID
=== RUN   TestRestAPI/Test_PUT_puppy_with_invalid_payload
=== RUN   TestRestAPI/Test_PUT_puppy_with_error
=== RUN   TestRestAPI/Test_PUT_with_invalid_id
=== RUN   TestRestAPI/Test_DELETE_puppy_by_ID
=== RUN   TestRestAPI/Test_DELETE_puppy_with_error_data
=== RUN   TestRestAPI/Test_DELETE_puppy_with_invalid_id
--- PASS: TestRestAPI (0.00s)
    --- PASS: TestRestAPI/Test_GET_Puppy_by_ID (0.00s)
    --- PASS: TestRestAPI/Test_GET_Puppy_with_invalid_input (0.00s)
    --- PASS: TestRestAPI/Test_GET_invalid_Puppy (0.00s)
    --- PASS: TestRestAPI/Test_POST_Puppy (0.00s)
    --- PASS: TestRestAPI/Test_POST_with_error_data (0.00s)
    --- PASS: TestRestAPI/Test_POST_with_decode_error (0.00s)
    --- PASS: TestRestAPI/Test_PUT_puppy_by_ID (0.00s)
    --- PASS: TestRestAPI/Test_PUT_puppy_with_invalid_payload (0.00s)
    --- PASS: TestRestAPI/Test_PUT_puppy_with_error (0.00s)
    --- PASS: TestRestAPI/Test_PUT_with_invalid_id (0.00s)
    --- PASS: TestRestAPI/Test_DELETE_puppy_by_ID (0.00s)
    --- PASS: TestRestAPI/Test_DELETE_puppy_with_error_data (0.00s)
    --- PASS: TestRestAPI/Test_DELETE_puppy_with_invalid_id (0.00s)
=== RUN   TestStorer
=== RUN   TestStorer/TestCreateNextPuppyWithCorrectID
=== RUN   TestStorer/TestCreatePuppySuccessfully
=== RUN   TestStorer/TestCreatePuppyWithValueLessThanZero
=== RUN   TestStorer/TestCreatePuppyWithWrongValue
=== RUN   TestStorer/TestDeleteNonExistentPuppy
=== RUN   TestStorer/TestDeletePuppy
=== RUN   TestStorer/TestReadNonExistentPuppy
=== RUN   TestStorer/TestReadPuppySuccessfully
=== RUN   TestStorer/TestUpdateNonExistentID
=== RUN   TestStorer/TestUpdatePuppySuccessfully
=== RUN   TestStorer/TestCreateNextPuppyWithCorrectID#01
=== RUN   TestStorer/TestCreatePuppySuccessfully#01
=== RUN   TestStorer/TestCreatePuppyWithValueLessThanZero#01
=== RUN   TestStorer/TestCreatePuppyWithWrongValue#01
=== RUN   TestStorer/TestDeleteNonExistentPuppy#01
=== RUN   TestStorer/TestDeletePuppy#01
=== RUN   TestStorer/TestReadNonExistentPuppy#01
=== RUN   TestStorer/TestReadPuppySuccessfully#01
=== RUN   TestStorer/TestUpdateNonExistentID#01
=== RUN   TestStorer/TestUpdatePuppySuccessfully#01
--- PASS: TestStorer (0.00s)
    --- PASS: TestStorer/TestCreateNextPuppyWithCorrectID (0.00s)
    --- PASS: TestStorer/TestCreatePuppySuccessfully (0.00s)
    --- PASS: TestStorer/TestCreatePuppyWithValueLessThanZero (0.00s)
    --- PASS: TestStorer/TestCreatePuppyWithWrongValue (0.00s)
    --- PASS: TestStorer/TestDeleteNonExistentPuppy (0.00s)
    --- PASS: TestStorer/TestDeletePuppy (0.00s)
    --- PASS: TestStorer/TestReadNonExistentPuppy (0.00s)
    --- PASS: TestStorer/TestReadPuppySuccessfully (0.00s)
    --- PASS: TestStorer/TestUpdateNonExistentID (0.00s)
    --- PASS: TestStorer/TestUpdatePuppySuccessfully (0.00s)
    --- PASS: TestStorer/TestCreateNextPuppyWithCorrectID#01 (0.00s)
    --- PASS: TestStorer/TestCreatePuppySuccessfully#01 (0.00s)
    --- PASS: TestStorer/TestCreatePuppyWithValueLessThanZero#01 (0.00s)
    --- PASS: TestStorer/TestCreatePuppyWithWrongValue#01 (0.00s)
    --- PASS: TestStorer/TestDeleteNonExistentPuppy#01 (0.00s)
    --- PASS: TestStorer/TestDeletePuppy#01 (0.00s)
    --- PASS: TestStorer/TestReadNonExistentPuppy#01 (0.00s)
    --- PASS: TestStorer/TestReadPuppySuccessfully#01 (0.00s)
    --- PASS: TestStorer/TestUpdateNonExistentID#01 (0.00s)
    --- PASS: TestStorer/TestUpdatePuppySuccessfully#01 (0.00s)
PASS
ok  	github.com/anz-bank/go-course/10_rest/willshen8/pkg/puppy	(cached)

Reply:
```shenw-JSS1256:willshen8 shenw$ go test -v ./...
=== RUN   TestParseCmdArgs
--- PASS: TestParseCmdArgs (0.00s)
=== RUN   TestParseCmdArgsWithError
--- PASS: TestParseCmdArgsWithError (0.00s)
=== RUN   TestShortFlag
--- PASS: TestShortFlag (0.00s)
=== RUN   TestParseCorrectPort
--- PASS: TestParseCorrectPort (0.00s)
=== RUN   TestParseWrongPort
--- PASS: TestParseWrongPort (0.00s)
=== RUN   TestCreateStorerWithCorrectArguments
=== RUN   TestCreateStorerWithCorrectArguments/map_storer
=== RUN   TestCreateStorerWithCorrectArguments/sync_storer
--- PASS: TestCreateStorerWithCorrectArguments (0.00s)
    --- PASS: TestCreateStorerWithCorrectArguments/map_storer (0.00s)
    --- PASS: TestCreateStorerWithCorrectArguments/sync_storer (0.00s)
=== RUN   TestCreateStorerWithWrongArguments
=== RUN   TestCreateStorerWithWrongArguments/wrong_storer
=== RUN   TestCreateStorerWithWrongArguments/empty_storer
--- PASS: TestCreateStorerWithWrongArguments (0.00s)
    --- PASS: TestCreateStorerWithWrongArguments/wrong_storer (0.00s)
    --- PASS: TestCreateStorerWithWrongArguments/empty_storer (0.00s)
=== RUN   TestSavePuppiesToStore
--- PASS: TestSavePuppiesToStore (0.00s)
=== RUN   TestSavePuppiesToStoreWithErr
--- PASS: TestSavePuppiesToStoreWithErr (0.00s)
=== RUN   TestParseJSONPuppiesSuccessfully
--- PASS: TestParseJSONPuppiesSuccessfully (0.00s)
=== RUN   TestParseBadJSONPuppies
--- PASS: TestParseBadJSONPuppies (0.00s)
=== RUN   TestParseBadJSONPuppiesWithErr
--- PASS: TestParseBadJSONPuppiesWithErr (0.00s)
=== RUN   TestParseEmptyJSONPuppies
--- PASS: TestParseEmptyJSONPuppies (0.00s)
=== RUN   TestMainHappyPath
unexpected
ok  	github.com/anz-bank/go-course/10_rest/willshen8/cmd/puppy-server	0.016s
=== RUN   TestUnknownError
--- PASS: TestUnknownError (0.00s)
=== RUN   TestRestAPI
=== RUN   TestRestAPI/Test_GET_Puppy_by_ID
=== RUN   TestRestAPI/Test_GET_Puppy_with_invalid_input
=== RUN   TestRestAPI/Test_GET_invalid_Puppy
=== RUN   TestRestAPI/Test_POST_Puppy
=== RUN   TestRestAPI/Test_POST_with_error_data
=== RUN   TestRestAPI/Test_POST_with_decode_error
=== RUN   TestRestAPI/Test_PUT_puppy_by_ID
=== RUN   TestRestAPI/Test_PUT_puppy_with_invalid_payload
=== RUN   TestRestAPI/Test_PUT_puppy_with_invalid_id
=== RUN   TestRestAPI/Test_PUT_with_invalid_id
=== RUN   TestRestAPI/Test_PUT_with_negative_puppy_value
=== RUN   TestRestAPI/Test_DELETE_puppy_by_ID
=== RUN   TestRestAPI/Test_DELETE_puppy_with_non_existent_id
=== RUN   TestRestAPI/Test_DELETE_puppy_with_invalid_id
--- PASS: TestRestAPI (0.00s)
    --- PASS: TestRestAPI/Test_GET_Puppy_by_ID (0.00s)
    --- PASS: TestRestAPI/Test_GET_Puppy_with_invalid_input (0.00s)
    --- PASS: TestRestAPI/Test_GET_invalid_Puppy (0.00s)
    --- PASS: TestRestAPI/Test_POST_Puppy (0.00s)
    --- PASS: TestRestAPI/Test_POST_with_error_data (0.00s)
    --- PASS: TestRestAPI/Test_POST_with_decode_error (0.00s)
    --- PASS: TestRestAPI/Test_PUT_puppy_by_ID (0.00s)
    --- PASS: TestRestAPI/Test_PUT_puppy_with_invalid_payload (0.00s)
    --- PASS: TestRestAPI/Test_PUT_puppy_with_invalid_id (0.00s)
    --- PASS: TestRestAPI/Test_PUT_with_invalid_id (0.00s)
    --- PASS: TestRestAPI/Test_PUT_with_negative_puppy_value (0.00s)
    --- PASS: TestRestAPI/Test_DELETE_puppy_by_ID (0.00s)
    --- PASS: TestRestAPI/Test_DELETE_puppy_with_non_existent_id (0.00s)
    --- PASS: TestRestAPI/Test_DELETE_puppy_with_invalid_id (0.00s)
=== RUN   TestStorer
=== RUN   TestStorer/TestCreateNextPuppyWithCorrectID
=== RUN   TestStorer/TestCreatePuppySuccessfully
=== RUN   TestStorer/TestCreatePuppyWithValueLessThanZero
=== RUN   TestStorer/TestCreatePuppyWithWrongValue
=== RUN   TestStorer/TestDeleteNonExistentPuppy
=== RUN   TestStorer/TestDeletePuppy
=== RUN   TestStorer/TestReadNonExistentPuppy
=== RUN   TestStorer/TestReadPuppySuccessfully
=== RUN   TestStorer/TestUpdateNonExistentID
=== RUN   TestStorer/TestUpdatePuppySuccessfully
=== RUN   TestStorer/TestUpdatePuppyWithNegativeValue
=== RUN   TestStorer/TestCreateNextPuppyWithCorrectID#01
=== RUN   TestStorer/TestCreatePuppySuccessfully#01
=== RUN   TestStorer/TestCreatePuppyWithValueLessThanZero#01
=== RUN   TestStorer/TestCreatePuppyWithWrongValue#01
=== RUN   TestStorer/TestDeleteNonExistentPuppy#01
=== RUN   TestStorer/TestDeletePuppy#01
=== RUN   TestStorer/TestReadNonExistentPuppy#01
=== RUN   TestStorer/TestReadPuppySuccessfully#01
=== RUN   TestStorer/TestUpdateNonExistentID#01
=== RUN   TestStorer/TestUpdatePuppySuccessfully#01
=== RUN   TestStorer/TestUpdatePuppyWithNegativeValue#01
--- PASS: TestStorer (0.00s)
    --- PASS: TestStorer/TestCreateNextPuppyWithCorrectID (0.00s)
    --- PASS: TestStorer/TestCreatePuppySuccessfully (0.00s)
    --- PASS: TestStorer/TestCreatePuppyWithValueLessThanZero (0.00s)
    --- PASS: TestStorer/TestCreatePuppyWithWrongValue (0.00s)
    --- PASS: TestStorer/TestDeleteNonExistentPuppy (0.00s)
    --- PASS: TestStorer/TestDeletePuppy (0.00s)
    --- PASS: TestStorer/TestReadNonExistentPuppy (0.00s)
    --- PASS: TestStorer/TestReadPuppySuccessfully (0.00s)
    --- PASS: TestStorer/TestUpdateNonExistentID (0.00s)
    --- PASS: TestStorer/TestUpdatePuppySuccessfully (0.00s)
    --- PASS: TestStorer/TestUpdatePuppyWithNegativeValue (0.00s)
    --- PASS: TestStorer/TestCreateNextPuppyWithCorrectID#01 (0.00s)
    --- PASS: TestStorer/TestCreatePuppySuccessfully#01 (0.00s)
    --- PASS: TestStorer/TestCreatePuppyWithValueLessThanZero#01 (0.00s)
    --- PASS: TestStorer/TestCreatePuppyWithWrongValue#01 (0.00s)
    --- PASS: TestStorer/TestDeleteNonExistentPuppy#01 (0.00s)
    --- PASS: TestStorer/TestDeletePuppy#01 (0.00s)
    --- PASS: TestStorer/TestReadNonExistentPuppy#01 (0.00s)
    --- PASS: TestStorer/TestReadPuppySuccessfully#01 (0.00s)
    --- PASS: TestStorer/TestUpdateNonExistentID#01 (0.00s)
    --- PASS: TestStorer/TestUpdatePuppySuccessfully#01 (0.00s)
    --- PASS: TestStorer/TestUpdatePuppyWithNegativeValue#01 (0.00s)
PASS
ok  	github.com/anz-bank/go-course/10_rest/willshen8/pkg/puppy	0.018s